File: backend/internal/services/notification_service.go
package services

import (
	"errors"

	"github.com/Kal-el21/booking-room-golang/backend/internal/models"
	"github.com/Kal-el21/booking-room-golang/backend/internal/repositories"
)

type NotificationService struct {
	notificationRepo *repositories.NotificationRepository
}

func NewNotificationService(notificationRepo *repositories.NotificationRepository) *NotificationService {
	return &NotificationService{
		notificationRepo: notificationRepo,
	}
}

// GetUserNotifications gets all notifications for current user
func (s *NotificationService) GetUserNotifications(userID uint, page, pageSize int) ([]models.Notification, int64, error) {
	if page < 1 {
		page = 1
	}
	if pageSize < 1 || pageSize > 100 {
		pageSize = 10
	}

	return s.notificationRepo.GetUserNotifications(userID, page, pageSize)
}

// GetUnreadCount gets unread notification count
func (s *NotificationService) GetUnreadCount(userID uint) (int64, error) {
	return s.notificationRepo.GetUnreadCount(userID)
}

// MarkAsRead marks a notification as read
func (s *NotificationService) MarkAsRead(id uint, userID uint) error {
	notification, err := s.notificationRepo.FindByID(id)
	if err != nil {
		return err
	}

	// Check permission
	if notification.UserID != userID {
		return errors.New("you don't have permission to mark this notification")
	}

	return s.notificationRepo.MarkAsRead(id)
}

// MarkAllAsRead marks all notifications as read for user
func (s *NotificationService) MarkAllAsRead(userID uint) error {
	return s.notificationRepo.MarkAllAsRead(userID)
}

// DeleteNotification deletes a notification
func (s *NotificationService) DeleteNotification(id uint, userID uint) error {
	notification, err := s.notificationRepo.FindByID(id)
	if err != nil {
		return err
	}

	// Check permission
	if notification.UserID != userID {
		return errors.New("you don't have permission to delete this notification")
	}

	return s.notificationRepo.Delete(id)
}

File: frontend/src/services/notification.service.ts
import api from './api';
import type { Notification, ApiResponse, PaginatedResponse } from '@/types';

const NOTIFICATION_PREFIX = '/api/v1/notifications';

export interface NotificationFilters {
  page?: number;
  page_size?: number;
}

export const notificationService = {
  // Get my notifications
  getNotifications: async (filters?: NotificationFilters): Promise<PaginatedResponse<Notification>> => {
    const params = new URLSearchParams();
    if (filters?.page) params.append('page', filters.page.toString());
    if (filters?.page_size) params.append('page_size', filters.page_size.toString());

    const response = await api.get<PaginatedResponse<Notification>>(
      `${NOTIFICATION_PREFIX}?${params.toString()}`
    );
    return response.data;
  },

  // Get unread count
  getUnreadCount: async (): Promise<number> => {
    const response = await api.get<ApiResponse<{ count: number }>>(
      `${NOTIFICATION_PREFIX}/unread-count`
    );
    return response.data.data.count;
  },

  // Mark as read
  markAsRead: async (id: number): Promise<void> => {
    await api.put(`${NOTIFICATION_PREFIX}/${id}/mark-as-read`);
  },

  // Mark all as read
  markAllAsRead: async (): Promise<void> => {
    await api.post(`${NOTIFICATION_PREFIX}/mark-all-as-read`);
  },

  // Delete notification
  deleteNotification: async (id: number): Promise<void> => {
    await api.delete(`${NOTIFICATION_PREFIX}/${id}`);
  },
};